docs/decisions/00NN-breaking-changes-guidance.md

```diff
@@ -0,0 +1,41 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: proposed
+contact: markwallace
+date: 2024-06-05
+deciders: sergeymenshykh, mbolan, rbarreto, dmytrostruk, westey
+consulted: 
+informed: 
+---
+
+# Guidance for Breaking Changes
+
+## Context and Problem Statement
+
+We must avoid breaking changes in .Net because of the well known [diamond dependency issue](https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/dependencies#diamond-dependencies) where breaking changes between different versions of the same package cause bugs and exceptions at run time.
+
+## Decision Drivers
+
+Breaking changes are only allowed under the following circumstances:
+
+- Updates to an experimental feature i.e. we have learnt something new and need to modify the design of an experimental feature.
+- When one of our dependencies introduces an unavoidable breaking change.
```

Comment:
I think we should be more explicit about how breaking changes are versioned when they are not on experimental features.
With reference to how Nuget does versioning https://learn.microsoft.com/en-us/nuget/concepts/package-versioning?tabs=semver20sort and specifically semantic versioning https://semver.org/spec/v2.0.0.html,
I think we should commit to keeping breaking changes to major version changes for non experimental features.

This also helps with the diamond problem, since packages that depend on us, would be able to specify a range of versions of SK that they are compatible with, within a single major version, e.g. > 1.12 and < 2.0.  That means that users can use all packages that target a single major version range together, without fear of failures.

Reply:
I don't think we will strictly follow semantic versioning as there may be minor breaking changes we need to make and we won't want to do a full version bump every time. We also want to keep the version numbers of our different language versions in sync.